File: src/domain/models/knowledge_entry.py
"""Domain models for the Knowledge Base bounded context.

KnowledgeCategory enumerates the four knowledge categories.
KnowledgePath is a frozen value object for RLM-addressable paths.
EntryMetadata holds verification and confidence metadata.
KnowledgeEntry is an entity identified by its KnowledgePath.
"""

from __future__ import annotations

import enum
from dataclasses import dataclass, field

from src.domain.models.errors import InvariantViolationError

_VALID_CATEGORIES = {"ddd", "tools", "conventions", "cross-tool"}


class KnowledgeCategory(enum.Enum):
    """Categories of knowledge entries in the Knowledge Base."""

    DDD = "ddd"
    TOOLS = "tools"
    CONVENTIONS = "conventions"
    CROSS_TOOL = "cross-tool"


@dataclass(frozen=True)
class KnowledgePath:
    """RLM-addressable path to a knowledge entry.

    Format: ``{category}/{topic}`` or ``{category}/{tool}/{subtopic}`` for tools.

    Attributes:
        raw: The original path string, e.g. "ddd/tactical-patterns".
    """

    raw: str

    def __post_init__(self) -> None:
        if not self.raw:
            msg = "Knowledge path must not be empty"
            raise InvariantViolationError(msg)
        if ".." in self.raw:
            msg = "Knowledge path must not contain path traversal (..)"
            raise InvariantViolationError(msg)
        segments = self.raw.split("/")
        if segments[0] not in _VALID_CATEGORIES:
            msg = (
                f"Knowledge path must start with a valid category "
                f"({', '.join(sorted(_VALID_CATEGORIES))}), got '{segments[0]}'"
            )
            raise InvariantViolationError(msg)

    @property
    def category(self) -> KnowledgeCategory:
        """Extract the category from the first path segment."""
        first = self.raw.split("/")[0]
        return KnowledgeCategory(first)

    @property
    def topic(self) -> str:
        """Extract the topic portion of the path.

        For tools paths (tools/tool-name/subtopic), returns "tool-name/subtopic".
        For other categories, returns the second segment.
        """
        segments = self.raw.split("/")
        if self.category == KnowledgeCategory.TOOLS:
            return "/".join(segments[1:])
        return segments[1]

    @property
    def tool(self) -> str | None:
        """Extract the tool name for TOOLS category paths, None otherwise."""
        if self.category != KnowledgeCategory.TOOLS:
            return None
        return self.raw.split("/")[1]

    @property
    def subtopic(self) -> str | None:
        """Extract the subtopic for TOOLS category paths, None otherwise."""
        if self.category != KnowledgeCategory.TOOLS:
            return None
        segments = self.raw.split("/")
        if len(segments) < 3:
            return None
        return segments[2]


@dataclass(frozen=True)
class EntryMetadata:
    """Verification and confidence metadata for a knowledge entry.

    Attributes:
        last_verified: ISO date when entry was last verified.
        verified_against: Tool/library version verified against.
        confidence: Confidence level (high, medium, low).
        deprecated: Whether this entry is deprecated.
        next_review_date: ISO date for next scheduled review.
        schema_version: Schema version of the entry format.
        source_urls: Authoritative source URLs for this entry.
    """

    last_verified: str | None = None
    verified_against: str | None = None
    confidence: str = "high"
    deprecated: bool = False
    next_review_date: str | None = None
    schema_version: str | None = None
    source_urls: tuple[str, ...] = ()


@dataclass
class KnowledgeEntry:
    """A knowledge entry in the Knowledge Base.

    Entity identity is based on the KnowledgePath -- two entries with the
    same path are considered equal regardless of content differences.

    Attributes:
        path: The RLM-addressable path for this entry.
        title: Human-readable title.
        content: The entry content (markdown or serialized TOML).
        metadata: Optional verification metadata.
        format: Content format, e.g. "markdown" or "toml".
    """

    path: KnowledgePath
    title: str
    content: str
    metadata: EntryMetadata | None = field(default=None)
    format: str = "markdown"

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, KnowledgeEntry):
            return NotImplemented
        return self.path.raw == other.path.raw

    def __hash__(self) -> int:
        return hash(self.path.raw)

File: internal/knowledge/domain/knowledge_entry.go
// Package domain provides the Knowledge bounded context's core domain model.
// It contains value objects for the knowledge base: categories, paths, metadata,
// entries, and drift detection.
package domain

import (
	"fmt"
	"strings"

	domainerrors "github.com/alto-cli/alto/internal/shared/domain/errors"
)

// KnowledgeCategory enumerates knowledge entry categories.
type KnowledgeCategory string

// Knowledge category constants.
const (
	CategoryDDD         KnowledgeCategory = "ddd"
	CategoryTools       KnowledgeCategory = "tools"
	CategoryConventions KnowledgeCategory = "conventions"
	CategoryCrossTool   KnowledgeCategory = "cross-tool"
)

var validCategories = map[string]KnowledgeCategory{
	"ddd":         CategoryDDD,
	"tools":       CategoryTools,
	"conventions": CategoryConventions,
	"cross-tool":  CategoryCrossTool,
}

// AllCategoryValues returns all valid category string values.
func AllCategoryValues() []string {
	return []string{
		string(CategoryDDD),
		string(CategoryTools),
		string(CategoryConventions),
		string(CategoryCrossTool),
	}
}

// ParseCategory parses a string into a KnowledgeCategory.
func ParseCategory(s string) (KnowledgeCategory, error) {
	cat, ok := validCategories[s]
	if !ok {
		return "", fmt.Errorf("invalid knowledge category %q: %w", s, domainerrors.ErrInvariantViolation)
	}
	return cat, nil
}

// KnowledgePath is an RLM-addressable path to a knowledge entry.
type KnowledgePath struct {
	raw string
}

// NewKnowledgePath creates a validated KnowledgePath.
func NewKnowledgePath(raw string) (KnowledgePath, error) {
	if raw == "" {
		return KnowledgePath{}, fmt.Errorf("knowledge path must not be empty: %w",
			domainerrors.ErrInvariantViolation)
	}
	if strings.Contains(raw, "..") {
		return KnowledgePath{}, fmt.Errorf("knowledge path must not contain path traversal (..): %w",
			domainerrors.ErrInvariantViolation)
	}
	segments := strings.Split(raw, "/")
	if _, ok := validCategories[segments[0]]; !ok {
		sorted := []string{"conventions", "cross-tool", "ddd", "tools"}
		return KnowledgePath{}, fmt.Errorf(
			"knowledge path must start with a valid category (%s), got '%s': %w",
			strings.Join(sorted, ", "), segments[0], domainerrors.ErrInvariantViolation)
	}
	if len(segments) < 2 || segments[1] == "" {
		return KnowledgePath{}, fmt.Errorf(
			"knowledge path requires category/topic format, got %q: %w",
			raw, domainerrors.ErrInvariantViolation)
	}
	return KnowledgePath{raw: raw}, nil
}

// Raw returns the raw path string.
func (p KnowledgePath) Raw() string { return p.raw }

// Category extracts the category from the first path segment.
func (p KnowledgePath) Category() KnowledgeCategory {
	first := strings.Split(p.raw, "/")[0]
	return validCategories[first]
}

// Topic extracts the topic portion of the path.
func (p KnowledgePath) Topic() string {
	segments := strings.Split(p.raw, "/")
	if p.Category() == CategoryTools {
		return strings.Join(segments[1:], "/")
	}
	return segments[1]
}

// Tool extracts the tool name for TOOLS category paths, nil otherwise.
func (p KnowledgePath) Tool() *string {
	if p.Category() != CategoryTools {
		return nil
	}
	segments := strings.Split(p.raw, "/")
	return &segments[1]
}

// Subtopic extracts the subtopic for TOOLS category paths, nil otherwise.
func (p KnowledgePath) Subtopic() *string {
	if p.Category() != CategoryTools {
		return nil
	}
	segments := strings.Split(p.raw, "/")
	if len(segments) < 3 {
		return nil
	}
	return &segments[2]
}

// EntryMetadata holds verification and confidence metadata for a knowledge entry.
type EntryMetadata struct {
	lastVerified    string
	verifiedAgainst string
	confidence      string
	nextReviewDate  string
	schemaVersion   string
	sourceURLs      []string
	deprecated      bool
}

// NewEntryMetadata creates an EntryMetadata value object.
func NewEntryMetadata(
	lastVerified, verifiedAgainst, confidence string,
	deprecated bool,
	nextReviewDate, schemaVersion string,
	sourceURLs []string,
) EntryMetadata {
	urls := make([]string, len(sourceURLs))
	copy(urls, sourceURLs)
	return EntryMetadata{
		lastVerified:    lastVerified,
		verifiedAgainst: verifiedAgainst,
		confidence:      confidence,
		deprecated:      deprecated,
		nextReviewDate:  nextReviewDate,
		schemaVersion:   schemaVersion,
		sourceURLs:      urls,
	}
}

// LastVerified returns when the entry was last verified.
func (m EntryMetadata) LastVerified() string { return m.lastVerified }

// VerifiedAgainst returns what the entry was verified against.
func (m EntryMetadata) VerifiedAgainst() string { return m.verifiedAgainst }

// Confidence returns the confidence level.
func (m EntryMetadata) Confidence() string { return m.confidence }

// Deprecated returns whether the entry is deprecated.
func (m EntryMetadata) Deprecated() bool { return m.deprecated }

// NextReviewDate returns when the entry should next be reviewed.
func (m EntryMetadata) NextReviewDate() string { return m.nextReviewDate }

// SchemaVersion returns the schema version.
func (m EntryMetadata) SchemaVersion() string { return m.schemaVersion }

// SourceURLs returns a defensive copy.
func (m EntryMetadata) SourceURLs() []string {
	out := make([]string, len(m.sourceURLs))
	copy(out, m.sourceURLs)
	return out
}

// KnowledgeEntry is an entity in the Knowledge Base, identified by its KnowledgePath.
type KnowledgeEntry struct {
	path     KnowledgePath
	title    string
	content  string
	metadata *EntryMetadata
	format   string
}

// NewKnowledgeEntry creates a KnowledgeEntry entity.
func NewKnowledgeEntry(path KnowledgePath, title, content string, metadata *EntryMetadata, format string) KnowledgeEntry {
	return KnowledgeEntry{
		path:     path,
		title:    title,
		content:  content,
		metadata: metadata,
		format:   format,
	}
}

// Path returns the knowledge entry path.
func (e KnowledgeEntry) Path() KnowledgePath { return e.path }

// Title returns the knowledge entry title.
func (e KnowledgeEntry) Title() string { return e.title }

// Content returns the knowledge entry content.
func (e KnowledgeEntry) Content() string { return e.content }

// Metadata returns the entry metadata.
func (e KnowledgeEntry) Metadata() *EntryMetadata { return e.metadata }

// Format returns the entry format.
func (e KnowledgeEntry) Format() string { return e.format }

// EqualByPath compares two entries by path identity.
func (e KnowledgeEntry) EqualByPath(other KnowledgeEntry) bool {
	return e.path.raw == other.path.raw
}
